[visionOS] Inform the client when Image Fullscreen is delegated to Quick Look

yoelhawa opened the pull request:
#### c63f556ac82e248edc81bd57bd430e4eb2012d98
<pre>
[visionOS] Inform the client when Image Fullscreen is delegated to Quick Look
<a href="https://bugs.webkit.org/show_bug.cgi?id=273970">https://bugs.webkit.org/show_bug.cgi?id=273970</a>
&lt;<a href="https://rdar.apple.com/126655604">rdar://126655604</a>&gt;

Reviewed by Aditya Keerthi.

Add a new UI delegate method to notify client when image fullscreen has
been delegated to Quick Look, and plumb the image dimensions to this
method.

* Source/WebKit/UIProcess/API/Cocoa/_WKFullscreenDelegate.h:
Add new UI delegate method.

* Source/WebKit/UIProcess/Cocoa/FullscreenClient.h:
* Source/WebKit/UIProcess/Cocoa/FullscreenClient.mm:
(WebKit::FullscreenClient::setDelegate):
(WebKit::FullscreenClient::didEnterFullscreen):
Call new UI delegate method.

* Source/WebKit/Shared/FullScreenMediaDetails.h:
* Source/WebKit/Shared/FullScreenMediaDetails.serialization.in:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::enterFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::enterFullScreen):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
Generalize name of variable `videoDimensions` to `mediaDimensions`.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
Expose some new public properties.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController isUsingQuickLook]):
(-[WKFullScreenWindowController imageDimensions]):
(-[WKFullScreenWindowController enterFullScreen:]):
Expose some new public properties, and generalize name of variable
`videoDimensions` to `mediaDimensions`.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
Add image dimensions to `mediaDetails` struct iff image fullscreen will
be delegated to Quick Look (to avoid interfering with downstream checks
of existence of dimensions).

Canonical link: <a href="https://commits.webkit.org/279728@main">https://commits.webkit.org/279728@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5445eca278c94c2a0e551e2f17a107681b98a29

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54162 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33550 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6699 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57438 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4886 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41071 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4782 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43862 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3263 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56258 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31818 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46898 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25009 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28603 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3035 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50309 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4423 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59031 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29363 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4528 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51303 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30537 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47010 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50643 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11832 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31504 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30317 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
```diff
@@ -206,6 +206,8 @@ struct UIEdgeInsets;
 
 - (void)_webView:(WKWebView *)webView didAdjustVisibilityWithSelectors:(NSArray<NSString *> *)selectors WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
 
+- (void)_webView:(WKWebView *)webView hasFullscreenedImageWithQuickLook:(CGSize)imageDimensions WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
`_webView:didFullscreenImageWithQuickLook:` maybe?
(`hasVideoInPictureInPicture` is a flag, so it makes sense there but not here)

etiennesegonzac commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1011,6 +1012,13 @@ - (void)enterFullScreen:(CGSize)videoDimensions
+
+#if PLATFORM(VISION) && ENABLE(QUICKLOOK_FULLSCREEN)
+        if (manager->isImageElement()) {
+            id webViewUIDelegate = [webView UIDelegate];
```
We need to check that the method exists before calling it
`if ([webViewUIDelegate respondsToSelector:@selector(_webView:hasFullscreenedImageWithQuickLook:)])`

etiennesegonzac commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -225,12 +225,14 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element,
     if (renderImage) {
+        auto* cachedImage = renderImage->cachedImage();
+        bool useQuickLookForFullscreen = cachedImage && cachedImage->image() && cachedImage->image()->shouldUseQuickLookForFullscreen();
+
```
We should lift the `errorOccured()` check too.

It might be easier to put everything in a IIFE so you can return early.
Something like
```c++
    if (renderImage) {
        mediaDetails = [&]() -> FullScreenMediaDetails {
            auto* cachedImage = renderImage->cachedImage();
            if (!cachedImage || cachedImage->errorOccurred())
                return { };

            RefPtr image = cachedImage->image();
            if (!image || !image->shouldUseQuickLookForFullscreen())
                return { };

            // ...
        }();
```

it'll simplify the rest of code too!

yoelhawa replied:
Good idea!

etiennesegonzac commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -245,9 +247,13 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element,
         };
 
         auto mimeType = emptyString();
+        FloatSize imageSize = { };
```
`FloatSize imageSize;` (it has a default constructor)

although this will probably become
```c++
            auto imageSize = image->size();
            auto mimeType = image->mimeType();
```
with the change above.

etiennesegonzac commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1011,6 +1012,13 @@ - (void)enterFullScreen:(CGSize)videoDimensions
         [CATransaction commit];
     }];
+
+#if PLATFORM(VISION) && ENABLE(QUICKLOOK_FULLSCREEN)
```
`#if ENABLE(QUICKLOOK_FULLSCREEN)` is enough
(I know we haven't been 100% consistent)

etiennesegonzac commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1011,6 +1012,13 @@ - (void)enterFullScreen:(CGSize)videoDimensions
     }];
+
+#if PLATFORM(VISION) && ENABLE(QUICKLOOK_FULLSCREEN)
+        if (manager->isImageElement()) {
```
Use `_isImageElement` instead (introduced recently).

etiennesegonzac commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
```diff
@@ -208,6 +208,7 @@ struct UIEdgeInsets;
 
 - (void)_webViewRecentlyAccessedGamepadsForTesting:(WKWebView *)webView WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
 - (void)_webViewStoppedAccessingGamepadsForTesting:(WKWebView *)webView WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
+- (void)_webView:(WKWebView *)webView didFullscreenImageWithQuickLook:(CGSize)imageDimensions WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
do we need `, visionos(WK_XROS_TBA)` like the one above?

etiennesegonzac commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -223,48 +223,48 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element,
-            };
-        }
-    }
+    mediaDetails = getMediaDetails();
```
it's a bit of a nit pick, but I would keep the `if (renderImage)` check outside of the lambda, and name it `getImageMediaDetails()`.
this way:
- we won't override any previously set mediaDetails when we don't have a RenderImage (if some code were to set it before this block in the future)
- it makes it clearer that there are other sources of media details like the code below

etiennesegonzac reviewed (commented):
A couple of comment but this is pretty much ready to go.
Flagging @pxlcoder for the official review!

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
```diff
@@ -208,6 +208,7 @@ struct UIEdgeInsets;
 
 - (void)_webViewRecentlyAccessedGamepadsForTesting:(WKWebView *)webView WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
 - (void)_webViewStoppedAccessingGamepadsForTesting:(WKWebView *)webView WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
+- (void)_webView:(WKWebView *)webView didFullscreenImageWithQuickLook:(CGSize)imageDimensions WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```
Nit: Let's add a newline before this method, since it's unrelated to the adjacent ones.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1011,6 +1012,14 @@ - (void)enterFullScreen:(CGSize)videoDimensions
     }];
+
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+        if (_isImageElement) {
```
Indentation is slightly off here.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1011,6 +1012,14 @@ - (void)enterFullScreen:(CGSize)videoDimensions
+        if (_isImageElement) {
+            id webViewUIDelegate = [webView UIDelegate];
+            if ([webViewUIDelegate respondsToSelector:@selector(_webView:didFullscreenImageWithQuickLook:)])
+                [webViewUIDelegate _webView:webView.get() didFullscreenImageWithQuickLook:mediaDimensions];
```
I'm not sure this is the right place to call this method. There are many places that can still cancel fullscreen following this call.

I would recommend following an approach similar to the existing `_webViewDidEnterFullscreen:` SPI.

See `UIDelegate::UIClient::didEnterFullscreen` / `WebFullScreenManagerProxy::didEnterFullScreen` / `FullscreenClient::didEnterFullscreen`. 

Updating `FullscreenClient::didEnterFullscreen` is probably the most correct. You should be able to read state off of `WKFullScreenWindowController` via `[m_webView fullScreenWindowController]`.

pxlcoder commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -225,45 +225,47 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element,
+            if (!renderImage)
+                return { };
+
+            auto* cachedImage = renderImage->cachedImage();
```
```suggestion
            RefPtr cachedImage = renderImage->cachedImage();
```

yoelhawa replied:
CachedImge doesn't support RefPtr

pxlcoder commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -225,45 +225,47 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element,
+                return { };
 
-            auto* image = renderImage->cachedImage()->image();
+            auto* image = cachedImage->image();
```
```suggestion
            RefPtr image = cachedImage->image();
```

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1011,6 +1025,11 @@ - (void)enterFullScreen:(CGSize)videoDimensions
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+    if (_isImageElement)
+        _imageDimensions = mediaDimensions;
+#endif
```
Can you move this higher up next to the other setup steps? Maybe after the `_rootViewController = adoptNS([[UIViewController alloc] init]);`.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/FullscreenClient.mm:
```diff
@@ -72,6 +76,14 @@
 {
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+    auto fullScreenController = [m_webView fullScreenWindowController];
+    if (fullScreenController) {
```
```suggestion
    if (auto fullScreenController = [m_webView fullScreenWindowController]) {
```

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/FullscreenClient.mm:
```diff
@@ -72,6 +76,14 @@
+        if (fullScreenController.isUsingQuickLook && m_delegateMethods.webViewDidFullscreenImageWithQuickLook)
+            [m_delegate.get() _webView:m_webView didFullscreenImageWithQuickLook:imageDimensions];
+    }
+#endif // ENABLE(QUICKLOOK_FULLSCREEN)
```
I would move this added logic to the end of the method. At the very least, I think it should not be called before `[m_webView willChangeValueForKey:@"fullscreenState"]`.

Commit:
[visionOS] Inform the client when Image Fullscreen is delegated to Quick Look
https://bugs.webkit.org/show_bug.cgi?id=273970
<rdar://126655604>

Reviewed by Aditya Keerthi.

Add a new UI delegate method to notify client when image fullscreen has
been delegated to Quick Look, and plumb the image dimensions to this
method.

* Source/WebKit/UIProcess/API/Cocoa/_WKFullscreenDelegate.h:
Add new UI delegate method.

* Source/WebKit/UIProcess/Cocoa/FullscreenClient.h:
* Source/WebKit/UIProcess/Cocoa/FullscreenClient.mm:
(WebKit::FullscreenClient::setDelegate):
(WebKit::FullscreenClient::didEnterFullscreen):
Call new UI delegate method.

* Source/WebKit/Shared/FullScreenMediaDetails.h:
* Source/WebKit/Shared/FullScreenMediaDetails.serialization.in:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::enterFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::enterFullScreen):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
Generalize name of variable `videoDimensions` to `mediaDimensions`.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
Expose some new public properties.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController isUsingQuickLook]):
(-[WKFullScreenWindowController imageDimensions]):
(-[WKFullScreenWindowController enterFullScreen:]):
Expose some new public properties, and generalize name of variable
`videoDimensions` to `mediaDimensions`.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
Add image dimensions to `mediaDetails` struct iff image fullscreen will
be delegated to Quick Look (to avoid interfering with downstream checks
of existence of dimensions).

Canonical link: https://commits.webkit.org/279728@main
